lib/src/utils.dart

```diff
@@ -0,0 +1,330 @@
+/// `false` otherwise.
+bool isPossibleBoolean(String input) {
+  final trimmedInput = input.trim();
+
+  switch (trimmedInput) {
+    case 'true':
+    case 'True':
+    case 'TRUE':
+    case 'false':
+    case 'False':
+    case 'FALSE':
+      return true;
+    default:
+      return false;
+  }
+}
+
+/// Returns `true` if [input] could be interpreted as a null value by `package:yaml`,
+/// `false` otherwise.
+bool isPossibleNull(String input) {
+  final trimmedInput = input.trim();
+
+  switch (trimmedInput) {
+    case '':
+    case 'null':
+    case 'Null':
+    case 'NULL':
+    case '~':
+      return true;
+    default:
+      return false;
+  }
+}
+
+/// Checks if [string] contains a YAML control character
+bool containsControlCharacter(String string) {
+  return (string.contains(':') ||
+      string.contains('!') ||
+      string.contains('&') ||
+      string.contains('*') ||
+      string.contains('-') ||
+      string.contains('[') ||
+      string.contains(']') ||
+      string.contains('#') ||
+      string.contains('|') ||
+      string.contains('>') ||
+      string.contains('@') ||
+      string.contains('`') ||
+      string.contains('"') ||
+      string.contains('{') ||
+      string.contains('}'));
+}
+
+/// Returns a safe string by ensuring that if [value] was meant to be a string, it
+/// will not be interpreted otherwise.
+String getSafeString(Object value) {
+  if (value is YamlNode) {
+    AssertionError(
+        'YamlNodes should not be passed directly into getSafeString!');
+  }
+
+  if (value is Map || value is List) {
+    AssertionError('Lists and Maps should not call getSafeString directly!');
+  }
+
+  if (value is String) {
+    var result = value.trim();
+
+    /// If it contains a dangerous character we want to wrap the result with single
+    /// quotes
+    if (containsControlCharacter(result) ||
+        isPossibleBoolean(result) ||
+        isPossibleNull(result)) {
+      /// But we need to escape the characters if they contain a single quote.
+      if (result.contains('\'')) {
+        result = result.replaceAll('\'', '\'\'');
+      }
+
+      return '\'$result\'';
+    }
+
+    return result;
+  }
+
+  return value.toString();
+}
+
+/// Returns [value] with the necessary formatting applied in a block context.
+///
+/// If [value] is a [YamlScalar], we try to respect its [style] parameter where
+/// possible. Certain cases make this impossible (e.g. a plain string scalar that
+/// starts with '>'), in which case we will produce [value] with default styling options.
+String getFlowScalar(Object value) {
+  if (value is Map || value is List) {
+    AssertionError('Only scalars can be passed into getScalarString');
+  }
+
+  if (value is YamlScalar) {
+    if (value.style == ScalarStyle.DOUBLE_QUOTED) {
+      return '"${value.value.toString()}"';
+    } else if (value.style == ScalarStyle.SINGLE_QUOTED) {
+      return '\'${value.value.toString()}\'';
+    }
```

Comment:
I touched on this in the comments of this method https://github.com/walnutdust/yaml_edit/pull/1/files/70516b815ac07eea5b047c50ef6923e29603bdb2#diff-1809444246e4db866a7cbd2139e00d63R94-R95, since it felt less like a TODO, but more of a problem that can't really be solved

Reply:
If the string contains values that need be escaped to work with quotation marks we should do such escaping, makes sense?

Also should we ever apply style if the scalar is a number?

Reply:
hm I just tested it and you're right - `package:yaml` parses numbers, booleans in the folded and literal styles as strings, so `ScalarStyle`s should only be applied if the scalar is a `String`.

And yes... I will have to escape all the values for the styles!